### ISSUE TITLE
AMQP Transport manage seneca host wrongly with multiple client

### ISSUE THREAD

**Issue by immanuel192** [seneca-contrib/seneca-amqp-transport#29]
Consider this code

```javascript
var clienta = require('seneca')()
    .use('seneca-amqp-transport')
    .client({
        type: 'amqp',
        url: 'amqp://gues:guest@localhost:5672',
        pin: 'module:a'
    });


var clientb = require('seneca')()
    .use('seneca-amqp-transport')
    .client({
        type: 'amqp',
        url: 'amqp://gues:guest@localhost:5672',
        pin: 'module:b'
    });

for (var i = 0; i < 100; i++) {
    var t1 = Date.now();
    clienta.act('module:a,role:create,foo:1,zed:' + i + ",t1:" + t1, function (err, ret) {
        var when = Date.now();
        var dt = when - ret.when;
        var dt2 = when - ret.t1;
        console.log("server a, with i = " + ret.zed + "; dt1 = " + dt + "; dt2 = " + dt2 + " ;  t1 = " + ret.when + " ; t2= " + when);
    });
    
    clientb.act('module:b,role:create,foo:1,zed:' + i + ",t1:" + t1, function (err, ret) {
        var when = Date.now();
        var dt = when - ret.when;
        var dt2 = when - ret.t1;
        console.log("server b, with i = " + ret.zed + "; dt1 = " + dt + "; dt2 = " + dt2 + " ;  t1 = " + ret.when + " ; t2= " + when);
    });
}
```

When running this code, you will see warning invalid_origin. All response from Server A to Client A will be invalid (invalid_origin). The issue is because, when initialize seneca-amqp-transport, maybe in somewhere, the plugin keep seneca host in wrongly way. I guess the issue is in client-hook.

That why only the last client will be alive, all other will be lost.

**Comment by nfantone** [comment]
@immanuel192 Could you please make proper use of [markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#code-and-syntax-highlighting) to clarify some of your snippets? It's somewhat hard to follow your examples.

**Comment by immanuel192** [comment]
Please check again

**Comment by immanuel192** [comment]
Here is the code on server
```javascript
seneca.add('foo:1', function (args, done) {
    var when = Date.now();
    var dt = when - args.t1;
    console.log("with i = " + args.zed + "; delta t = " + dt + "; t1 = " + args.t1 + " ; t2= " + when);
    done(null, { t1: args.t1, zed: args.zed, bar: args.zed + 1, when: when })
});
```

**Comment by nfantone** [comment]
@immanuel192 Well, your example is a very odd one. Are you playing around with Seneca or is this a real use case?

For starters, I could point out a couple of things that I'd do differently:

1. Don't create more than _one_ instance of `seneca` and its plugins. Instead, require it once and reuse that.
2. There's no need to declare multiple clients. You could just define a "multi-pin" client, which is exactly what you want. Take a look at the [examples](https://github.com/seneca-contrib/seneca-amqp-transport/blob/develop/examples/multipin-client.js).

Lastly, please bear in mind that pins _must_ match on either side of the communication.

**Comment by immanuel192** [comment]
Hi @nfantone 

Please check my plugin in here: https://bitbucket.org/trungdang_apium/microservices_servicebase/src
This is plugin for my real project. 

My idea is:
- Create a ServiceBase, as a base of my microservices
- Redis, Consul will be integrated into ServiceBase. 
- Each service of my microservices will :
  + Use RabbitMQ to communicate to each other
  + Each kind of this service (for example: UserService, ProductService, ...) will have their own pin. For example: UserService will have their pin as module:user
  + We will not know exactly how many kind of services in system as system will grow up. 
  + Each kind of this service (such as UserService) can be started as multi- instances (multi Docker containers).

Because of all reasons above, I assume that my code will not know exactly the pins / multi-pin to use when registering the seneca client. That why I try to create new instance of seneca when I need a client for specific module.

**Comment by nfantone** [comment]
@immanuel192 From what I can gather, your scenario is a very typical microservice use case. I may be misunderstanding something, but it seems to me that nothing special prevents you from doing what you want with Seneca + this plugin.

However, I fail to understand the relationship between the "plugin" you are developing the issue you have raised in your first post.

> Each kind of this service (such as UserService) can be started as multi- instances

Great! That can be easily done. It's automatically handled for you. Just spawn multiple instances of the same script, using the same sets of pins.

> I assume that my code will not know exactly the pins / multi-pin to use when registering the seneca client. 

I can't think of a single scenario where you "cannot" know the pins during registration of clients or listeners. Pins are mandatory. You cannot "not know" them. They may be dynamically generated, yes. But, in some way or another, you end up building/yielding them.

>  That why I try to create new instance of seneca when I need a client for specific module.

Similarly, I'm struggling to follow your reasoning here. Could you elaborate on this? Why do you need multiple instances of `seneca` on the same microservice script? I see no advantages on doing so.

**Comment by immanuel192** [comment]
Hi @nfantone 

Thanks for your comment. 

I solved the issue by  just init seneca one time and reuse it.

**Comment by nfantone** [comment]
Hey, @immanuel192, you're welcome. Glad to be of help.
